File: java/amp-commanding/src/main/java/amp/commanding/DefaultCommandReceiver.java
package amp.commanding;


import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import amp.messaging.IMessageHandler;
import amp.messaging.IMessageProcessor;
import amp.messaging.MessageException;
import amp.messaging.MessageReceiver;
import cmf.bus.Envelope;
import cmf.bus.IEnvelopeReceiver;


/**
 * Created with IntelliJ IDEA.
 * User: jar349
 * Date: 5/1/13
 */
public class DefaultCommandReceiver extends MessageReceiver implements ICommandReceiver {

    public DefaultCommandReceiver(IEnvelopeReceiver envelopeReceiver) {
        super(envelopeReceiver);
    }

    public DefaultCommandReceiver(IEnvelopeReceiver envelopeReceiver, List<IMessageProcessor> processorChain) {
        super(envelopeReceiver, processorChain);
    }

    @Override
    public <TCOMMAND> void onCommandReceived(ICommandHandler<TCOMMAND> handler) throws MessageException, IllegalArgumentException {
    	onMessageReceived(new CommandMessageHandler<TCOMMAND>(handler));
    }
    
    private static class CommandMessageHandler<TCOMMAND> implements IMessageHandler<TCOMMAND>{
    	private static final Logger LOG = LoggerFactory.getLogger(CommandMessageHandler.class);

        private final ICommandHandler<TCOMMAND> _handler;
  
        public CommandMessageHandler(ICommandHandler<TCOMMAND> handler){
            _handler = handler;
        }

        public Class<TCOMMAND> getMessageType(){
            return _handler.getCommandType(); 
        }

        @Override
		public Object handle(TCOMMAND message, Map<String, String> headers){
            _handler.handle(message, headers);
            return null;
        }

        @Override
        public Object handleFailed(Envelope env, Exception ex){
        	LOG.warn("Failed to handle envelope.", ex);
            return null;
        }
    }
}

File: dotnet/source/amp.commanding/DefaultCommandReceiver.cs
using System;
using System.Collections.Generic;
using amp.messaging;
using cmf.bus;
using Common.Logging;

namespace amp.commanding
{
    public class DefaultCommandReceiver : MessageReceiver, ICommandReceiver
    {
        public DefaultCommandReceiver(IEnvelopeReceiver envelopeReceiver) 
            : base(envelopeReceiver)
        {
        }

        public DefaultCommandReceiver(IEnvelopeReceiver envelopeReceiver, List<IMessageProcessor> processingChain) 
            : base(envelopeReceiver, processingChain)
        {
        }

        public void ReceiveCommand<TCommand>(Action<TCommand, IDictionary<string, string>> handler) where TCommand : class
        {
            base.ReceiveMessage(handler);
        }

        public void ReceiveCommand(ICommandHandler handler)
        {
            base.ReceiveMessage(new CommandMessageHandler(handler));
        }

        private class CommandMessageHandler : IMessageHandler
        {
            private static readonly ILog Log = LogManager.GetLogger(typeof(CommandMessageHandler));

            private readonly ICommandHandler _handler;
      
            public CommandMessageHandler(ICommandHandler handler)
            {
                _handler = handler;
            }

            public string Topic
            {
                get { return _handler.CommandType.FullName; }
            }

            public object Handle(object message, IDictionary<string, string> headers)
            {
                return _handler.Handle(message, headers);
            }

            public object HandleFailed(Envelope env, Exception ex)
            {
                Log.Warn("Failed to handle envelope.", ex);
                return null;
            }
        }
    }
}
